### PR TITLE
Users can submit asynchronous, long running queries via the API

### PR DESCRIPTION
**Pull request by MDunitz**
# Need
- As a user of the Query Service,
- I want to submit queries to the API asynchronously,
- so that I can issue queries that take longer than HTTP timeouts.

# Approach
- long-running-query post endpoint accepts a query string and returns a job_id
- long-running-query/{job_id} get endpoint returns the status of the long query job and if its complete, a link to the s3 object containing the results 
- after a user posts to the long-running-query endpoint the query is added to a queue and eventually processed by a lambda which writes the query and the results of the query to a new object in the query-service s3 bucket

# Tests
- [x] unit tests
- [x] functional tests
- [x] manual testing of presigned url permissions

# Todo
- [ ] cron job for deleting rows from job_status table
- [ ] more restrictive permissions?
- [ ] prevent people from writing entire db to file (Select * from files;)?

### CODE REVIEW

File: query/test/functional/test_endpoints.py [hunk 1]
```diff
@@ -45,3 +87,5 @@ def _make_request(self, description, verb, url, expected_status=None, **options)
             print(response.content.decode('utf8'))
 
         return response.content
+
+
```
Comment (mweiden) [inline query/test/functional/test_endpoints.py]: rm extra white space

File: query/test/unit/create_long_query_lambda/test_create_long_query.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+from lambdas.long_query.create_long_query import query_db_and_put_results_in_s3
+from test.unit import QueryTestCaseUsingMockAWS
+
+
+class TestCreateLongQuery(QueryTestCaseUsingMockAWS):
+    def setUp(self):
+        super().setUp()
+        self.uuid = '26f0424a-fdce-455f-ac2e-f8f5619c6eda'
+        self.query = "SELECT * FROM files limit 10;"
+        self.mock_event = {'Records': [{'messageId': 'AAAAA',
```
Comment (mweiden) [inline query/test/unit/create_long_query_lambda/test_create_long_query.py]: I'd make these fields static members of `TestCreateLongQuery`. Then you don't need to define `setUp` and `tearDown`.

File: query/test/unit/test_database.py [hunk 3]
```diff
@@ -127,6 +128,25 @@ def test_list(tables: Tables, num_intersecting_tables: int):
             with self.db.transaction() as (_, tables):
                 test_list(tables, 0)
 
+    def test_job_status__creation(self):
```
Comment (mweiden) [inline query/test/unit/test_database.py]: extra `_`

File: query/test/unit/test_database.py [hunk 4]
```diff
@@ -127,6 +128,25 @@ def test_list(tables: Tables, num_intersecting_tables: int):
...
+        uuid = uuid4()
+        with self.db.transaction() as (cursor, tables):
+            row_count = tables.job_status.insert(uuid)
+            assert row_count == 1
+            job = tables.job_status.select(uuid)
+            assert job['status'] == 'CREATED'
+            job = tables.job_status.select_from_write_db(uuid)
+            assert job['status'] == 'CREATED'
+
+    def test_job_status__status_update(self):
```
Comment (mweiden) [inline query/test/unit/test_database.py]: here too

File: terraform/modules/query-service-infra/long_query_lambda.tf [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+resource "aws_lambda_function" "query_create_long_query_lambda" {
+  function_name    = "query-create-long-query-${var.deployment_stage}"
+  s3_bucket        = "${aws_s3_bucket.query-service.id}"
+  s3_key           = "${var.deployment_stage}/lambda_deployments/create_long_query/create_long_query.zip"
+  role             = "arn:aws:iam::${local.account_id}:role/query-create-long-query-${var.deployment_stage}"
+  handler          = "app.create_long_query"
+  runtime          = "python3.6"
+  memory_size      = 960
+  timeout          = 900
```
Comment (mweiden) [inline terraform/modules/query-service-infra/long_query_lambda.tf]: Should the timeout need to be this long given it is async?

File: chalice/.chalice/config.json [hunk 6]
```diff
@@ -7,7 +7,9 @@
       "environment_variables": {
         "DEPLOYMENT_STAGE": "dev",
         "API_HOST": "query.dev.data.humancellatlas.org",
-        "LOAD_DATA_QUEUE_URL": "https://sqs.us-east-1.amazonaws.com/861229788715/dcp-query-data-input-queue-dev"
+        "LOAD_DATA_QUEUE_URL": "https://sqs.us-east-1.amazonaws.com/861229788715/dcp-query-data-input-queue-dev",
+        "LONG_QUERY_QUEUE_URL": "https://sqs.us-east-1.amazonaws.com/861229788715/dcp-query-long-query-queue-dev",
+        "AWS_ACCOUNT_ID": "861229788715"
```
[the remark this answers is not in the document]
Comment (kislyuk) [reply chalice/.chalice/config.json]: Thanks for the heads up. Will work with @MDunitz to refactor this

[the remark this answers is not in the document]
Comment (kislyuk) [reply chalice/.chalice/config.json]: Strategy: Let's use the parameter store to save these.

File: terraform/modules/query-service-infra/long_query_lambda.tf (hunk 5 above)
Comment (mweiden) [reply terraform/modules/query-service-infra/long_query_lambda.tf]: ah, I see, it blocks on the connection to the db. Got it.

File: config/query-api.yml [hunk 7]
```diff
@@ -88,6 +88,57 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  /v1/long-running-query:
+    post:
+      summary: Endpoint to kickoff longer running query of hca metadata
+      description: Receives a query that may take longer than 30 seconds to complete. Returns a job id which can be used in a get request to retrieve the query results
+      operationId: query.lambdas.api_server.v1.endpoints.create_long_query
+      tags:
+        - All
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply config/query-api.yml]: Discussed in person

File: config/query-api.yml [hunk 8]
```diff
@@ -88,6 +88,57 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  /v1/long-running-query:
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply config/query-api.yml]: 👍

File: create_long_query_lambda/.gitignore [hunk 9]
```diff
@@ -0,0 +1,3 @@
+target/
+create_long_query.zip
+vendor/
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply create_long_query_lambda/.gitignore]: 👍

File: create_long_query_lambda/Makefile [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+	cd target && zip -r ../$(ZIP_FILE) *
+
+stage: build
+	aws s3 cp $(ZIP_FILE) s3://$(BUCKET)/$(STAGED_FILE_KEY)
+
+deploy: stage
+	aws lambda update-function-code --function-name query-create-long-query-$(DEPLOYMENT_STAGE) --s3-bucket $(BUCKET) --s3-key $(STAGED_FILE_KEY)
+
+clobber: ;
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply create_long_query_lambda/Makefile]: removed

File: query/lambdas/api_server/v1/endpoints.py [hunk 11]
```diff
@@ -13,6 +14,8 @@
 
 db = PostgresDatabase(Config.serve_database_uri)
 sqs_client = boto3.client('sqs')
+s3_client = boto3.client('s3')
+
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply query/lambdas/api_server/v1/endpoints.py]: 👍

File: create_long_query_lambda/app.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import json
+
+from query.lambdas.long_query.create_long_query import query_db_and_put_results_in_s3
+from query.lib.common.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+def create_long_query(event, context):
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply create_long_query_lambda/app.py]: 👍

File: query/lambdas/api_server/v1/endpoints.py [hunk 13]
```diff
@@ -28,20 +31,49 @@ def query(query_string):
     return {'query': query_string, "results": formatted_query_results}, requests.codes.ok
 
 
+@return_exceptions_as_http_errors
+def create_long_query(query_string):
+    query_string = json.loads(query_string, strict=False)
+    uuid = uuid4()
+    with db.transaction() as (_, tables):
+        tables.job_status.insert(uuid)
+    queue_url = Config.long_query_queue_url
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply query/lambdas/api_server/v1/endpoints.py]: 👍

File: query/lambdas/long_query/create_long_query.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        data = json.dumps({"query": query_string, "results": formatted_query_results}, default=str)
+        deployment_stage = Config.deployment_stage
+
+        account_id = Config.account_id
+        s3.put_object(ACL='authenticated-read',
+                      Body=data,
+                      Bucket=f"query-service-{account_id}",
+                      Key=f"{deployment_stage}/query_results/{job_id}")
+        update_job_status(job_id, 'COMPLETE')
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply query/lambdas/long_query/create_long_query.py]: 👍

File: query/test/functional/test_endpoints.py (hunk 1 above)
Comment (MDunitz) [reply query/test/functional/test_endpoints.py]: 👍

File: query/test/unit/test_database.py (hunk 3 above)
Comment (MDunitz) [reply query/test/unit/test_database.py]: 👍

File: query/test/unit/test_database.py (hunk 4 above)
Comment (MDunitz) [reply query/test/unit/test_database.py]: 👍

File: terraform/modules/query-service-infra/long_query_lambda.tf (hunk 5 above)
Comment (MDunitz) [reply terraform/modules/query-service-infra/long_query_lambda.tf]: 👍

File: query/test/unit/create_long_query_lambda/test_create_long_query.py (hunk 2 above)
Comment (MDunitz) [reply query/test/unit/create_long_query_lambda/test_create_long_query.py]: 👍

File: chalice/.chalice/config.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (MDunitz) [reply chalice/.chalice/config.json]: I think that should be a separate PR. I'll add a ticket

[the remark this answers is not in the document]
Comment (mweiden) [reply chalice/.chalice/config.json]: @MDunitz sounds good, ty

File: query/lambdas/async_query/create_async_query.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+from query.lib.common.logging import get_logger
+from query.lib.common.formatting import format_query_results
+from query.lib.config import Config
+from query.lib.db.database import PostgresDatabase
+from query.lib.logger import logger
+
+db = PostgresDatabase(Config.serve_database_uri)
+s3 = boto3.client('s3')
+
+# logger = get_logger('query.lambdas.long_query.create_long_query')
```
Comment (mweiden) [inline query/lambdas/async_query/create_async_query.py]: rm this and extra newlines

File: query/test/functional/test_endpoints.py [hunk 16]
```diff
@@ -30,6 +30,48 @@ def test_query(self):
         self.assertEqual(json.loads(response)['query'], query)
         self.assertGreaterEqual(json.loads(response)['results'][0]['count'], 0)
 
+    def test_create_long_running_query(self):
```
Comment (mweiden) [inline query/test/functional/test_endpoints.py]: You should probably change this to reflect your name change to `create_async_query`. There's a bunch more occurrences of `long_query` and `long_running_query` in the codebase. I'd do a text search and make all of them consistent.

**Comment by mweiden** [comment]
@MDunitz 🎈🚢 👏

**Pull request merged**